lesson30/src/lib.rs

```diff
@@ -0,0 +1,585 @@
+    num::{NonZeroU128, NonZeroU32},
+};
+
+type AccountId = String;
+type Money = u32;
+type NonZeroMoney = NonZeroU32;
+type Err = String;
+type OpId = NonZeroU128;
+
+#[derive(Debug, Default, PartialEq, Eq)]
+pub struct Account {
+    pub balance: Money,
+}
+
+#[derive(Debug, Clone)]
+pub enum Operation {
+    /**
+     * Перед выполнением любых операций по счёту, его необходимо создать.
+     * Если счёт с именем Х существует, то создание нового счёта с именем Х - ошибка
+     */
+    Create(AccountId), //регистрация счёта
+
+    /**Пополнение увеличивает количество денег на счете на указанную сумму.
+     * Пополнение на ноль денежных единиц - ошибка.
+     */
+    Charge(NonZeroMoney), //пополнение
+    /**
+     * Снятие уменьшает количество денег на счете на указанную сумму.
+     * Снятие нуля денежных единиц - ошибка.
+     * Попытка снять больше чем есть на счете - ошибка.
+     */
+    Withdraw(NonZeroMoney), //снятие
+                            //перевол реализован через сумму операций Withdraw + Charge
+}
+
+pub trait OpsStorage {
+    fn transact(
+        &mut self,
+        ops: impl Iterator<Item = (AccountId, Operation)>,
+    ) -> Result<impl Iterator<Item = (OpId, AccountId, &Operation)>, Err>;
+
+    fn persist(&mut self, account_id: AccountId, op: Operation) -> Result<(OpId, &Operation), Err> {
+        self.transact(vec![(account_id, op)].into_iter())
+            .and_then(|mut iter| {
+                iter.next()
+                    .map(|(opt_id, _, op)| (opt_id, op))
+                    .ok_or("a transaction should return at least one operation".to_owned())
+            })
+    }
+    //should be O(N), where N - account ops
+    fn get_ops<'a, 'b>(
+        &'a self,
+        account_id: &'b AccountId,
+    ) -> Result<impl Iterator<Item = (OpId, &'a Operation)>, Err>;
+    //should be O(M), where M - all ops
+    fn get_history(&self) -> Result<impl Iterator<Item = (OpId, (AccountId, &Operation))>, Err>;
+}
+
+pub trait State {
+    fn transact<'a, 'b>(
+        &'a mut self,
+        ops: impl Iterator<Item = (AccountId, &'b Operation)>,
+    ) -> Result<impl Iterator<Item = (AccountId, &'a Account)>, Err>;
+
+    fn update<'a, 'b>(
+        &'a mut self,
+        account_id: AccountId,
+        op: &'b Operation,
+    ) -> Result<&'a Account, Err> {
+        self.transact(vec![(account_id, op)].into_iter())
+            .and_then(|mut iter| {
+                iter.next()
+                    .map(|(_, acc)| acc)
+                    .ok_or("a transaction should return at least one account".to_owned())
+            })
+    }
+    //should be O(N), where N - account ops
+    fn get_balance(&self, account_id: &AccountId) -> Result<&Account, Err>;
+}
+
+pub struct Bank<T: OpsStorage, S: State> {
+    storage: T,
+    state: S,
+}
+
+impl<T: OpsStorage, S: State> Bank<T, S> {
+    pub fn new(storage: T, state: S) -> Bank<T, S> {
+        Bank {
+            storage: storage,
+            state: state,
+        }
```

Comment:
```suggestion
        Bank {
            storage,
            state,
        }
```

Если имена совпадают, то писать явно так необязательно